{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fbde3611",
   "metadata": {},
   "source": [
    "## Assignment 3:  Convolutional networks for TF ChIP-seq data\n",
    "\n",
    "In this assignment you will build on the convolutional networks we looked at in class and work on ChIP-seq data for four transcription factors in arabidopsis.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9313bb24-1ac1-4263-914f-cef4af557c3e",
   "metadata": {},
   "source": [
    "### Part 1: Data Preparation\n",
    "\n",
    "In this assignment you will work with ChIP-seq for four arabidopsis transcription factors:  AGL16, GRF1, AMS, and MYB3R4.  The peaks that represent their binding sites in the arabidopsis genome are available in the following links:\n",
    "\n",
    "* AGL16 ([bed file](https://biobigdata.nju.edu.cn/ChIPHub_download/arabidopsis_thaliana/SRP187795/hammock/AGL16.target.all.bed.gz))\n",
    "* GRF1 ([bed file](https://biobigdata.nju.edu.cn/ChIPHub_download/arabidopsis_thaliana/SRP002566/hammock/SRX021610.peak.all.bed.gz))\n",
    "* AMS ([bed file](https://biobigdata.nju.edu.cn/ChIPHub_download/arabidopsis_thaliana/SRP188198/hammock/SRX5507861.peak.all.bed.gz))\n",
    "* MYB3R4 ([bed file](https://biobigdata.nju.edu.cn/ChIPHub_download/arabidopsis_thaliana/SRP244735/hammock/SRX7616358.peak.all.bed.gz)).\n",
    "\n",
    "These files are in [bed format](https://en.wikipedia.org/wiki/BED_(file_format)), and contain the information on the genomic locations where the ChIP-seq peaks have been detected.  The linked wikipedia article provides the information you need about the format of these files.  Your task is to extract sequences of length 500 centered at the location of each peak, which you will provide as input to the convolutional network you train.  \n",
    "\n",
    "The other piece of information you need is the genomic sequence for arabidopsis.  This is available from the [Ensembl plants arabidopsis portal](https://plants.ensembl.org/Arabidopsis_thaliana/Info/Index).  In that page click on \"Download DNA sequence (FASTA)\", and the first five files provide the sequences for the five arabidopsis chromosomes.\n",
    "For reference, we computed the result for AGL16 (link is in the assignment page in Canvas).\n",
    "\n",
    "Your final data preparation task is to prepare a labeled dataset with positive examples that correspond to the peak sequences.  As negative examples, use random permutations of the positive examples.  Create one permutation from each positive example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "46b8394c",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[WinError 3] The system cannot find the path specified: '/mnt/c/Users/hunte/Desktop/CS525'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[1], line 10\u001b[0m\n\u001b[0;32m      8\u001b[0m path\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/mnt/c/Users/hunte/Desktop/CS525\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m      9\u001b[0m \u001b[38;5;66;03m#path=r'C:\\Users\\hunte\\Desktop\\CS525'\u001b[39;00m\n\u001b[1;32m---> 10\u001b[0m \u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mchdir\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpath\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     12\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mwarnings\u001b[39;00m\n\u001b[0;32m     13\u001b[0m warnings\u001b[38;5;241m.\u001b[39mfilterwarnings(action\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124monce\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[1;31mFileNotFoundError\u001b[0m: [WinError 3] The system cannot find the path specified: '/mnt/c/Users/hunte/Desktop/CS525'"
     ]
    }
   ],
   "source": [
    "from numpy import random\n",
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "path='/mnt/c/Users/hunte/Desktop/CS525'\n",
    "#path=r'C:\\Users\\hunte\\Desktop\\CS525'\n",
    "os.chdir(path)\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(action='once')\n",
    "\n",
    "\n",
    "import pybedtools #import Bedtool\n",
    "\n",
    "usingLinux=False\n",
    "\n",
    "#path=r'C:\\\\Users\\\\hunte\\Desktop\\\\CS525'\n",
    "#path='/mnt/c/Users/hunte/Desktop/CS525'\n",
    "#os.chdir(path)\n",
    "#Loop through BED file\n",
    "\n",
    "#bedFiles = ['AGL16.target.all.bed','SRX021610.peak.all.bed','SRX5507861.peak.all.bed','SRX7616358.peak.all.bed']\n",
    "\n",
    "bedFiles = ['RNAPosProm.txt','RNANegProm.txt']\n",
    "\n",
    "#Read bed line by line, taking first three numbers\n",
    "\n",
    "posSeqs = []\n",
    "#This will be a x by y array, where x is the number of bed files with peak data and y is the number of sequences in a single bed file\n",
    "#Each element will be a string of 100 characters representing a positive peak for a given sample\n",
    "\n",
    "#def getSeq(chrom,start,end):\n",
    "\n",
    "#    return\n",
    "\n",
    "#ColumnNames = ['AGL16_Pos','AGL16_Neg','GRF1_Pos','GRF1_Neg','AMS_Pos','AMS_Neg','MYR3R4_Pos','MYR3R4_Neg']\n",
    "\n",
    "#ColumnNames = ['Ac_Pos','Ac_Neg','Me_Pos','Me_Neg']\n",
    "\n",
    "ColumnNames = ['DiffSeqs','ControlSeqs']\n",
    "\n",
    "tfNames = ['H3K9Me3','H3K27Ac']\n",
    "\n",
    "fileNum=0\n",
    "for file in bedFiles:\n",
    "    filePos = []\n",
    "    bedFile = open(file,'r')\n",
    "    bedLines = bedFile.readlines()\n",
    "    #temp = 0\n",
    "    fastaFile = open(tfNames[fileNum]+'_HO_peaks.fasta','w')\n",
    "    for line in bedLines:\n",
    "        fields = line.split()\n",
    "        midPeak = round((int(fields[1])+int(fields[2]))/2)\n",
    "\n",
    "        #SWITCH BACK FOR CS525 ASSIGNMENT\n",
    "\n",
    "        peakLow = midPeak-750\n",
    "        peakHigh = midPeak+750\n",
    "        #if(peakLow < 0):\n",
    "            #peakLow = 0\n",
    "            #peakHigh = 500\n",
    "        #position = str(fields[0][3]) + \"\\t\" + str(fields[1]) + \"\\t\" + str(fields[2])\n",
    "        #position = str(fields[0][3]) + \"\\t\" + str(peakLow) + \"\\t\" + str(peakHigh)\n",
    "\n",
    "        ####### CHANGE TO PREVIOUS FOR CS525\n",
    "\n",
    "        position = str(fields[0]) + \"\\t\" + str(peakLow) + \"\\t\" + str(peakHigh)\n",
    "\n",
    "        #Need to account for possibility of extending beyond edge of chromosomes\n",
    "\n",
    "        #print(position)\n",
    "        posBed = pybedtools.BedTool(position,from_string=True)\n",
    "        try:\n",
    "            posSeq = posBed.sequence(fi=\"AedesGenome68.fa\")\n",
    "        except:\n",
    "            continue\n",
    "        #print(posSeq)\n",
    "        test = open(posSeq.seqfn)\n",
    "        print(test)\n",
    "        #print(open(posSeq.seqfn).read().split()[1])\n",
    "        a = test.read().split()[1]\n",
    "        #a = test.read()#.split()[1]\n",
    "        #print(a)\n",
    "        #posSeq.seqfn.close()\n",
    "        labelString = \">\" + tfNames[fileNum] + ',' + fields[0] + \",\" + str(peakLow) + ',' + str(peakHigh) + ',500'\n",
    "        filePos.append(a)\n",
    "        test.close()\n",
    "        fastaFile.write(labelString+'\\n')\n",
    "        fastaFile.write(a+'\\n')\n",
    "        #temp=temp+1\n",
    "        #if(temp > 10):\n",
    "        #    break\n",
    "    fastaFile.close()\n",
    "    bedFile.close()\n",
    "    posSeqs.append(filePos)\n",
    "    fileNum=fileNum+1\n",
    "\n",
    "\n",
    "#Get positive example from appopriate chromosome as a string\n",
    "\n",
    "#pos_example =  \n",
    "\n",
    "#Generate permutation of sequence to use as a negative example\n",
    "\n",
    "negSeqs = []\n",
    "\n",
    "for pos in posSeqs:\n",
    "    fileNeg = []\n",
    "    for seq in pos:\n",
    "        #print(seq)\n",
    "        #neg_example = random.permutation(seq)\n",
    "        #print(seq)\n",
    "        #print(random.shuffle(list(seq)))\n",
    "        #neg_example = str(random.shuffle(list(seq)))\n",
    "        negSeq = list(seq)\n",
    "        #print(seq)\n",
    "        #print(negSeq)\n",
    "        random.seed(2024)\n",
    "        random.shuffle(negSeq)\n",
    "        #print(negSeq)\n",
    "        negSeq = \"\".join(negSeq)\n",
    "        #print(negSeq)\n",
    "        #neg_example = seq\n",
    "        #Can simplify after debugging\n",
    "        fileNeg.append(negSeq)\n",
    "    negSeqs.append(fileNeg)\n",
    "\n",
    "#Add both to data frame\n",
    "\n",
    "#ColumnNames = ['Ac_Pos','Ac_Neg','Me_Pos','Me_Neg']\n",
    "ColumnNames = ['DiffRNA','ControlRNA']\n",
    "\n",
    "#samplesFrame = pd.DataFrame([posSeqs[0],negSeqs[0],posSeqs[1],negSeqs[1],\n",
    "#                             posSeqs[2,negSeqs[2],posSeqs[3],negSeqs[3]]],columns=ColumnNames)\n",
    "\n",
    "#for i in seq(4):\n",
    "\n",
    "\n",
    "#negSeqs.shape()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "627a774b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "posSave = open(\"PosExamplesRNAProm.pkl\", 'wb') \n",
    "pickle.dump(posSeqs, posSave)\n",
    "negSave = open(\"NegExamplesRNAProm.pkl\", 'wb') \n",
    "pickle.dump(negSeqs, negSave)\n",
    "\n",
    "negSave.close()\n",
    "posSave.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2113f2e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''import pickle\n",
    "\n",
    "testFile=\"PosExamples.pkl\"\n",
    "#posSave = open(\"Assignment3DataRead.pkl\", 'wb')\n",
    "posSeqs=pd.read_pickle(testFile)\n",
    "\n",
    "testFile=\"NegExamples.pkl\"\n",
    "negSeqs=pd.read_pickle(testFile)\n",
    "\n",
    "\n",
    "#testFile=\"PosExamples.pkl\"\n",
    "#pickle.dump(dataset, dataSave)\n",
    "#dataSave.close()'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7ccdb871",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''import itertools\n",
    "\n",
    "DFNames = ['Transcription Factor','Event_ID','Seq','Bound']\n",
    "\n",
    "#tfNames = ['AGL16','GRF1','AMS','MYR3R4']\n",
    "\n",
    "tfNames = ['H3K9Me3','H3K27Ac']\n",
    "\n",
    "\n",
    "for i in range(4):\n",
    "    #trainFile = open(tfNames[i]+'_HO.seq','w')\n",
    "    \n",
    "    seqNum = len(posSeqs[i])\n",
    "    events = list(range(2*(seqNum-1)))\n",
    "    eventNums = [\"seq_\"+str(event)+\"_peak\" for event in events]\n",
    "    TF_name = list(itertools.repeat(tfNames[i],2*(seqNum-1)))\n",
    "    bound = list(itertools.repeat(1,seqNum-1))+list(itertools.repeat(0,seqNum-1))\n",
    "    allSeqs = posSeqs[i][:] + negSeqs[i][:]\n",
    "    #allSeqs.append(negSeqs[i])\n",
    "    #allSeqs = posSeqs[i].append(negSeqs[i])\n",
    "    #print(len(TF_name))\n",
    "    #print(len(allSeqs))\n",
    "    #print(len(eventNums))\n",
    "    #print(len(bound))\n",
    "\n",
    "\n",
    "    tempFrame = pd.DataFrame(zip(*(TF_name,eventNums,allSeqs,bound)),columns=DFNames)\n",
    "    #tempFrame.reset_index(drop=True,inplace=True)\n",
    "    tempFrame.to_csv(tfNames[i]+'_HO.seq',sep='\\t',index=False)\n",
    "    #trainFile.write(labelString+'\\n')\n",
    "    #trainFile.write(a+'\\n')\n",
    "    #trainFile.close()'''\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bea7438",
   "metadata": {},
   "source": [
    "### Part 2:  \n",
    "\n",
    "As discussed in class, deeper networks with multiple layers of convolution *can* improve a network's performance.  Your task here is to extend the implementation provided in class to have three layers of convolution.  In addition, implement early stopping based on performance on the validation set.\n",
    "Finally, train each network two or three times, and choose the best performing network (based on the performance on the validation set) as the one to evaluate on the test set.\n",
    "In your experiments, set aside 20% of the data for testing, 20% for validation, and 60% for training.\n",
    "Compare the accuracy of your network to that of a one layer CNN.  Accuracy should be measured using the area under the ROC curve.  In the next part of the assignment you will get to tune its parameters to try and improve its performance.\n",
    "\n",
    "In this part, you can focus on two out of the four datasets provided above.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ChipNN",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
